x-pack/plugin/deprecation/src/main/java/org/elasticsearch/xpack/deprecation/DeprecationInfoAction.java

```diff
@@ -146,43 +147,57 @@ private static Map<DeprecationIssue, List<String>> getMergedIssuesToNodesMap(
+            "templates",
+            "ilm_policies"
+        );
+        static final Set<String> RESOURCE_CHECKER_FIELD_NAMES = Set.of("index_settings", "data_streams", "templates", "ilm_policies");
         private final List<DeprecationIssue> clusterSettingsIssues;
         private final List<DeprecationIssue> nodeSettingsIssues;
-        private final Map<String, List<DeprecationIssue>> indexSettingsIssues;
-        private final Map<String, List<DeprecationIssue>> dataStreamIssues;
+        private final Map<String, Map<String, List<DeprecationIssue>>> resourceDeprecationIssues;
         private final Map<String, List<DeprecationIssue>> pluginSettingsIssues;
 
         public Response(StreamInput in) throws IOException {
             super(in);
             clusterSettingsIssues = in.readCollectionAsList(DeprecationIssue::new);
             nodeSettingsIssues = in.readCollectionAsList(DeprecationIssue::new);
-            indexSettingsIssues = in.readMapOfLists(DeprecationIssue::new);
-            if (in.getTransportVersion().onOrAfter(TransportVersions.DATA_STREAM_INDEX_VERSION_DEPRECATION_CHECK)) {
-                dataStreamIssues = in.readMapOfLists(DeprecationIssue::new);
-            } else {
-                dataStreamIssues = Map.of();
+            Map<String, Map<String, List<DeprecationIssue>>> mutableResourceDeprecations = in.getTransportVersion()
+                .before(TransportVersions.RESOURCE_DEPRECATION_CHECKS) ? new HashMap<>() : Map.of();
+            if (in.getTransportVersion().before(TransportVersions.RESOURCE_DEPRECATION_CHECKS)) {
+                mutableResourceDeprecations.put(IndexDeprecationChecker.NAME, in.readMapOfLists(DeprecationIssue::new));
             }
+            if (in.getTransportVersion().before(TransportVersions.DATA_STREAM_INDEX_VERSION_DEPRECATION_CHECK)) {
+                mutableResourceDeprecations.put(DataStreamDeprecationChecker.NAME, Map.of());
+            } else if (in.getTransportVersion()
+                .between(TransportVersions.DATA_STREAM_INDEX_VERSION_DEPRECATION_CHECK, TransportVersions.RESOURCE_DEPRECATION_CHECKS)) {
+                    mutableResourceDeprecations.put(DataStreamDeprecationChecker.NAME, in.readMapOfLists(DeprecationIssue::new));
+                }
             if (in.getTransportVersion().before(TransportVersions.V_7_11_0)) {
                 List<DeprecationIssue> mlIssues = in.readCollectionAsList(DeprecationIssue::new);
                 pluginSettingsIssues = new HashMap<>();
                 pluginSettingsIssues.put("ml_settings", mlIssues);
             } else {
                 pluginSettingsIssues = in.readMapOfLists(DeprecationIssue::new);
             }
+            if (in.getTransportVersion().onOrAfter(TransportVersions.RESOURCE_DEPRECATION_CHECKS)) {
+                mutableResourceDeprecations = in.readMap(in2 -> in2.readMapOfLists(DeprecationIssue::new));
+            }
+            resourceDeprecationIssues = mutableResourceDeprecations;
```

Comment:
Should we create an immutable version of the mutable version here, so that we don't leak the mutability?